infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServerCache.java

```diff
@@ -62,159 +78,160 @@ public class SequencerServerCache implements AutoCloseable {
                 .orElseGet(HashMap::new);
+
+        this.versionMap = MicroMeterUtils
+                .gauge(WINDOW_SIZE_NAME, new TreeMap<Long, Set<ConflictTxStream>>(), TreeMap::size)
+                .orElseGet(TreeMap::new);
     }
 
     /**
-     * Returns the value associated with the {@code key} in this cache,
-     * or {@code null} if there is no cached value for the {@code key}.
-     *
-     * @param conflictKey conflict stream
-     * @return global address
+     * Returns the global address associated with the conflict key in this cache,
+     * or {@code Address.NON_ADDRESS} if there is no such cached address.
+     * @param conflictKey The conflict key.
+     * @return The global address associated with this hashed conflict key, if present.
      */
-    public Long get(ConflictTxStream conflictKey) {
-        return conflictKeys.getOrDefault(conflictKey, Address.NON_ADDRESS);
+    public long get(@NonNull ConflictTxStream conflictKey) {
+        return conflictKeyMap.getOrDefault(conflictKey, Address.NON_ADDRESS);
     }
 
     /**
-     * The first address in the priority queue.
+     * Returns the first (smallest) address in the cache.
+     * @return The smallest address currently in the cache.
      */
     public long firstAddress() {
-        if (cacheEntries.isEmpty()) {
+        if (versionMap.isEmpty()) {
             return Address.NOT_FOUND;
         }
-        return cacheEntries.peek().txVersion;
-    }
 
-  /**
-   * Invalidate the records with the minAddress. It could be one or multiple records
-   *
-   * @return the number of entries has been invalidated and removed from the cache.
-   */
-  private int invalidateSmallestTxVersion() {
-    ConflictTxStream firstEntry = cacheEntries.peek();
-    if (cacheEntries.size() == 0) {
-      return 0;
+        return versionMap.firstKey();
     }
 
-    int numEntries = 0;
-
-    while (firstAddress() == firstEntry.txVersion) {
-      if (log.isTraceEnabled()) {
-        log.trace(
-            "invalidateSmallestTxVersion: items evicted {} min address {}",
-            numEntries,
-            firstAddress());
-      }
-      ConflictTxStream entry = cacheEntries.poll();
-      conflictKeys.remove(entry);
-      numEntries++;
+    /**
+     * Evict the records with the smallest address in the cache.
+     * It could be one or multiple records.
+     *
+     * @return The number of entries that have been evicted from the cache.
+     */
+    private int evictSmallestTxVersion() {
+      final long firstAddress = firstAddress();
+      final Set<ConflictTxStream> entriesToDelete = versionMap.remove(firstAddress);
+      final int numDeletedEntries = entriesToDelete.size();
+      log.trace("evictSmallestTxVersion: items evicted {} min address {}", numDeletedEntries, firstAddress);
+      entriesToDelete.forEach(conflictKeyMap::remove);
+      maxConflictWildcard = Math.max(maxConflictWildcard, firstAddress);
```

Comment:
This caller logic that uses  `long maxConflictWildcard = cache.getMaxConflictWildcard();` assumes this is the first address in the window, but it's actually not in the window, its an evicted address. This makes the ` if (txSnapshotTimestamp.getSequence() < maxConflictWildcard)` invalid, no?

Reply:
Can you clarify what is meant here? From what I can tell, the caller does not assume that `maxConflictWildcard` is the first address in the window. This is stated explicitly in the comment above the if condition you reference.